test/e2e/storage/testsuites/provisioning.go

```diff
@@ -323,17 +320,25 @@ func (t StorageClassTest) TestDynamicProvisioning() *v1.PersistentVolume {
 	var err error
 	if class != nil {
 		framework.ExpectEqual(*claim.Spec.StorageClassName, class.Name)
-		ginkgo.By("creating a StorageClass " + class.Name)
-		_, err = client.StorageV1().StorageClasses().Create(context.TODO(), class, metav1.CreateOptions{})
-		// The "should provision storage with snapshot data source" test already has created the class.
-		// TODO: make class creation optional and remove the IsAlreadyExists exception
-		framework.ExpectEqual(err == nil || apierrors.IsAlreadyExists(err), true)
+
 		class, err = client.StorageV1().StorageClasses().Get(context.TODO(), class.Name, metav1.GetOptions{})
-		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting storage class %s", class.Name)
-			framework.ExpectNoError(client.StorageV1().StorageClasses().Delete(context.TODO(), class.Name, metav1.DeleteOptions{}))
-		}()
+		if err == nil {
+			// skip storageclass creation if it already exists, this case happens when
+			// preparePVCDataSourceForProvisioning is called first
+			ginkgo.By("Storage class " + class.Name + " is already created, skipping creation.")
+		} else {
+			ginkgo.By("creating a StorageClass " + class.Name)
+			_, err = client.StorageV1().StorageClasses().Create(context.TODO(), class, metav1.CreateOptions{})
+			// The "should provision storage with snapshot data source" test already has created the class.
+			// TODO: make class creation optional and remove the IsAlreadyExists exception
```

Comment:
I think it would be cleaner if this function is more explicit about which tests already create the StorageClass ahead of time and which do not. Or change all tests to create the StorageClass outside of TestDynamicProvisioning so that we're consistent across all test cases.

Reply:
hey @msau42, @jingxu97 pointed out one nice observation which is that if this e2e test wasn't testing the implicit usage of the default storage class then we might need one or we might modify this one too, another strategy that I was implementing was to have each go routine create its own resources (its own storageclass, its own pvc, its own injector, etc) but I'm not sure if the test is about sharing the storageclass that's created before the go routines run.

Another thing that I wanted to ask was if creation of resources should be detached from properties in the testcase, if so then I can modify the signature to accept more parameters to control the creation like this:

```go
func (t StorageClassTest) TestDynamicProvisioning(shouldCreateStorageClass, shouldCreatePVC bool) *v1.PersistentVolume {
```

Reply:
I think we don't need each go routine to have its own storageclass. The main purpose of the test case creating the storageclass, is if it needs to modify any fields such as bindingMode or allowVolumeExpansion. But generally the StorageClass is considered a shared resource.

Agree that we should have test cases that explicitly test default storageclass vs non-default. For tests run under storage/testsuites, we don't test the default storageclass. For tests one level above "test/e2e/storage" we have some default storageclass cases, although it may be worth considering adding another config for testsuites just to use the default storageclass (should be considered separate from this issue).

I think this function needs to continue to handle both default and non-default storageclasses. My main suggestion was to simplify the logic so that in either case, the StorageClass is created outside of this function.